### PR TITLE
8301599: Serial: Refactor nested closures in DefNewGeneration

### PR DESCRIPTION
**Pull request by albertnetymk**
Simple refactoring broken into two commits:

1. Rename `KeepAliveClosure`.
2. Extract and move closures to cpp file.

Test: hotspot_gc

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8301599](https://bugs.openjdk.org/browse/JDK-8301599): Serial: Refactor nested closures in DefNewGeneration

### Reviewers
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**)
 * [Ivan Walulya](https://openjdk.org/census#iwalulya) (@walulyai - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12359/head:pull/12359` \
`$ git checkout pull/12359`

Update a local copy of the PR: \
`$ git checkout pull/12359` \
`$ git pull https://git.openjdk.org/jdk pull/12359/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12359`

View PR using the GUI difftool: \
`$ git pr show -t 12359`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12359.diff">https://git.openjdk.org/jdk/pull/12359.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/gc/serial/defNewGeneration.cpp [hunk 1]
```diff
@@ -63,43 +63,75 @@
...
-}
+    Devirtualizer::do_oop(_cl, p);
 
-void DefNewGeneration::FastKeepAliveClosure::do_oop(oop* p)       { DefNewGeneration::FastKeepAliveClosure::do_oop_work(p); }
-void DefNewGeneration::FastKeepAliveClosure::do_oop(narrowOop* p) { DefNewGeneration::FastKeepAliveClosure::do_oop_work(p); }
+    // Optimized for Defnew generation if it's the youngest generation:
+    // we set a younger_gen card if we have an older->youngest
+    // generation pointer.
+    oop obj = RawAccess<IS_NOT_NULL>::oop_load(p);
+    if ((cast_from_oop<HeapWord*>(obj) < _boundary) && GenCollectedHeap::heap()->is_in_reserved(p)) {
```
Comment (tschatzl) [inline src/hotspot/share/gc/serial/defNewGeneration.cpp:101]: I'm wondering whether the second part of the condition can ever be false. From what I can see, the `KeepAliveClosure` is only ever used during reference processing (afaict, may be wrong), and I can't think of a reason why the `Reference` would ever be outside the heap.
Feel free to ignore for this change, I can clean this up.

File: src/hotspot/share/gc/serial/defNewGeneration.cpp [hunk 2]
```diff
@@ -63,43 +63,75 @@
...
+  Generation* _young_gen;
+public:
+  IsAliveClosure(Generation* young_gen) : _young_gen(young_gen) {
+    assert(_young_gen->kind() == Generation::DefNew,
+        "Expected the young generation here");
+  }
 
-// Methods of protected closure types.
+  bool do_object_b(oop p) {
+    return cast_from_oop<HeapWord*>(p) >= _young_gen->reserved().end() || p->is_forwarded();
```
Comment (tschatzl) [inline src/hotspot/share/gc/serial/defNewGeneration.cpp:75]: Another micro-optimization for another day, `_young_gen->reserved().end()` is constant across the use (i.e. avoid the dereference for `_young_gen->reserved()` all the time).

**Review by tschatzl — APPROVED** [review]
Lgtm.

**Review by walulyai — APPROVED** [review]

**Comment by albertnetymk** [comment]
Thanks for the review.

/integrate

**Pull request closed without merging**